jobrunner/record_stats.py

```diff
@@ -28,24 +28,29 @@
 )
 """
 
-_conn = None
 
+def get_connection(readonly=True):
+    if readonly:
+        db = config.METRICS_FILE + "?mode=ro"
+    else:
+        db = config.METRICS_FILE
 
-def ensure_metrics_db():
-    global _conn
-    _conn = database.get_connection(config.METRICS_FILE)
-    _conn.execute("PRAGMA journal_mode = WAL")
-    _conn.execute(DDL)
+    return database.get_connection(db, initial=["PRAGMA journal_mode = WAL", DDL])
 
 
 def read_job_metrics(job_id, **metrics):
-    if _conn is None:
-        ensure_metrics_db()
+    try:
+        raw_metrics = (
+            get_connection()
+            .execute(
+                "SELECT metrics FROM jobs WHERE id = ?",
+                (job_id,),
+            )
+            .fetchone()
+        )
+    except database.CannotOpenReadonlyDBFile:
+        raw_metrics = None
```

Comment:
There's another possible failure mode here in that the database file could exist but not have the relevant tables created yet. Rather than raising `CannotOpenReadonlyDBFile` from `get_connection()` I'd be more inclined to catch general "SQLite barfed" errors here i.e. `read_job_metrics` is a best-effort function that just returns None if it can't read for any reason.